File: vigenere.py
import sys


def read_file(file_name):
    with open(file_name, 'r') as file_name_arg:
        file_str = file_name_arg.read().replace('\n', '').replace(' ', '')

    # Only need the alphabetic letter, then make them lowercase.
    return ''.join(filter(str.isalpha, file_str)).lower()


def print_to_console(text, limit):
    count = 0

    for char in text:
        print(char, end='')
        # Put a newline when we print up to the limit.
        if (count + 1) % limit == 0:
            print()
        count += 1
    print()


def making_plaintext(plaintext_arg):
    plaintext_arg = read_file(plaintext_arg)

    while len(plaintext_arg) < 512:
        plaintext_arg += 'x'

    if len(plaintext_arg) > 512:
        plaintext_arg = plaintext_arg[:512]

    return plaintext_arg


def modify_key_string(key_string):
    key_init_len = len(key_string)
    key_tracker = 0

    while len(key_string) < 512:
        key_string += key_string[key_tracker]
        key_tracker += 1

        if key_tracker >= key_init_len:
            key_tracker = 0

    return key_string


def building_ciphertext(key_string, plaintext_string):
    cipher_string = ""

    for plaintext_char, key_char in zip(plaintext_string, key_string):
        plaintext_num = ord(plaintext_char) - ord('a')
        key_num = ord(key_char) - ord('a')
        combined_num = (plaintext_num + key_num) % 26
        cipher_string += chr(combined_num + ord('a'))

    return cipher_string


def main():
    if len(sys.argv) != 3:
        print('Usage: python3 vigenere.py keyFile plaintextFile', file=sys.stderr)
        sys.exit(1)

    key_str = read_file(sys.argv[1])

    print()
    print()

    print('Vigenere Key:')
    print()

    print_to_console(key_str, 80)

    print()
    print()

    plaintext_str = making_plaintext(sys.argv[2])

    print('Plaintext:')
    print()

    print_to_console(plaintext_str, 80)

    print()
    print()

    key_str = modify_key_string(key_str)

    cipher_str = building_ciphertext(key_str, plaintext_str)

    print('Ciphertext:')
    print()

    print_to_console(cipher_str, 80)


if __name__ == "__main__":
    main()

File: vigenere.go
package main

import (
    "fmt"
    "log"
    "os"
    "regexp"
    "strings"
)

func printToStdout(sFile string, limit int) {
    for i := 0; i < len(sFile); i++ {
        fmt.Print(string(sFile[i]))

        if (i + 1) % limit == 0 {
            fmt.Println()
        }
    }
    fmt.Println()
}

func readFile(pathname string) string {
    b, err := os.ReadFile(pathname)

    if err != nil {
        fmt.Print(err)
    }

    str := string(b)

    reg, err := regexp.Compile("[^A-Za-z]+")
    if err != nil {
        fmt.Print(err)
    }
    newStr := reg.ReplaceAllString(str, "")
    newStr = strings.ToLower(newStr)
    return newStr
}

func createPlainText(plaintextArg string) string {
    var sb strings.Builder
    sb.WriteString(plaintextArg)

    for sb.Len() < 512 {
        sb.WriteRune('x')
    }

    if sb.Len() > 512 {
        return sb.String()[:512]
    } else {
        return sb.String()
    }
}

func createKeyStr(keyStr string) string {
    var sb strings.Builder
    sb.WriteString(keyStr)
    initialKeyLen := sb.Len()
    keyTracker := 0

    for sb.Len() < 512 {
        strTemp := keyStr[keyTracker]
        sb.WriteString(string(strTemp))
        keyTracker += 1

        if keyTracker >= initialKeyLen {
            keyTracker = 0
        }
    }
    return sb.String()
}

func createCiphertext(keyStr string, plaintextStr string) string {
    var sb strings.Builder
    for i := 0; i < len(plaintextStr); i++ {
        plainTextDec := plaintextStr[i] - 'a'
        keyDec := keyStr[i] - 'a'
        addDec := (plainTextDec + keyDec) % 26
        cipherChar := string(addDec + 'a')
        sb.WriteString(cipherChar)
    }
    return sb.String()
}

func main() {
    if len(os.Args) != 3 {
        log.Fatal("Usage: ./vigenere keyFile plaintextFile")
        os.Exit(1)
    }

    keyArg := os.Args[1]
    plaintextArg := os.Args[2]

    fmt.Println()
    fmt.Println()

    keyStr := readFile(keyArg)

    fmt.Println("Vigenere Key:")
    fmt.Println()

    printToStdout(keyStr, 80)

    fmt.Println()
    fmt.Println()

    plaintextStr := readFile(plaintextArg)

    plaintextStr = createPlainText(plaintextStr)

    fmt.Println("Plaintext:")
    fmt.Println()

    printToStdout(plaintextStr, 80)

    fmt.Println()
    fmt.Println()

    keyStr = createKeyStr(keyStr)

    cipherStr := createCiphertext(keyStr, plaintextStr)

    fmt.Println("Ciphertext:")
    fmt.Println()

    printToStdout(cipherStr, 80)
}
